### PR TITLE
feat(lsp): dynamically generate list title in response_to_list

### PR DESCRIPTION
**Pull request by gpanders**
This gives quickfix/location lists created by handlers which use
`response_to_list` (textDocument/documentSymbol and workspace/symbol by
default) the ability to set a more useful list title. This commit gives
lists created for documentSymbol a title of the form:

    Symbols in <filename>

and lists for workspace/symbol a title of the form:

    Symbols matching '<query>'

These are more informative than a standard "Language Server" list title
and can help disambiguate results when users have multiple quickfix
lists that they cycle through with `:colder` and `:cnewer`.

### CODE REVIEW

**Review by mfussenegger — COMMENTED** [review]
Could we set a static context in addition or so?

Probably a niche use-case, but I have a custom quickfixfunc that turns entries like `jdt://contents/java.xml/[...]/ListDatatypeValidator.class` into `package.name::ClassName: [Class] ListDatatypeValidator` and I'm currently using the title as a marker for whether it should process the items or fallback to the default.

Change itself would otherwise look good to me.

**Comment by gpanders** [comment]
> Could we set a static context in addition or so?

Do you mean passing `ctx` as the quickfix context in `setqflist()`?

```diff
-          title = 'Language Server';
+          title = title_fn(ctx);
+          context = ctx;
           items = map_result(result, ctx.bufnr);
```

**Comment by mfussenegger** [comment]
> > Could we set a static context in addition or so?
> 
> Do you mean passing `ctx` as the quickfix context in `setqflist()`?
> 
> ```diff
> -          title = 'Language Server';
> +          title = title_fn(ctx);
> +          context = ctx;
>            items = map_result(result, ctx.bufnr);
> ```

That would be one way to do it, yes. For my use case it would work.

**Review by mfussenegger — DISMISSED** [review]

**Comment by gpanders** [comment]
The most robust way to get the word being referenced from `vim.lsp.buf.references()` is to capture it when references() is called and forward it to the handler. The LSP server itself doesn't communicate to the client what the word is (only the cursor position), and the cursor may have moved between the request and response.

To facilitate this, add a new `user_data` field to client requests that is not sent to the LSP server, but is simply forwarded to the client's response handler.

(This is a simple change, but it still merits its own commit, so I'll separate this PR into two commits).

**Comment by mjlbach** [comment]
Is this worth it? The implementation is a bit ugly IMO.

**Comment by gpanders** [comment]
> Is this worth it? 

That's up to you.

**Review by mjlbach — APPROVED** [review]

**Review by mfussenegger — APPROVED** [review]

File: runtime/lua/vim/lsp/buf.lua [hunk 1]
```diff
@@ -30,7 +30,7 @@ end
 ---
 ---@param method (string) LSP method name
 ---@param params (optional, table) Parameters to send to the server
----@param handler (optional, functionnil) See |lsp-handler|. Follows |lsp-handler-resolution|
+---@param handler (optional, function) See |lsp-handler|. Follows |lsp-handler-resolution|
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/buf.lua:33]: I assume these changes are left over, I'm fine with them ofc but it should be collected into a bigger style cleanup PR

**Pull request merged**